### PR TITLE
Email verification

### PR DESCRIPTION
**Pull request by Jachin-Manda**
- Add email verification after the user registration
- Add Sendgrid as a service for sending  emails

### CODE REVIEW

File: app/base/routes.py [hunk 1]
```diff
@@ -27,6 +28,52 @@ def route_errors(error):
...
+    try:
+        email = confirm_token(token)
+        user = User.query.filter_by(email=email).first()
+        if user.is_verified:
+            flash("Your email is already verified.", "success")
+        else:
+            user.is_verified = True
+            user.date_confirmed = datetime.utcnow
+            db.session.commit()
+            flash("Your email has been verified.", "success")
```
Comment (sonlinux) [inline app/base/routes.py:42]: perhaps just
"Your email has been successfully verified."

File: app/base/routes.py [hunk 2]
```diff
@@ -27,6 +28,52 @@ def route_errors(error):
     return render_template("errors/{}.html".format(error))
 
 
+@blueprint.route("/verify_email/<token>")
+def verify_email(token):
+    try:
+        email = confirm_token(token)
+        user = User.query.filter_by(email=email).first()
+        if user.is_verified:
+            flash("Your email is already verified.", "success")
```
Comment (sonlinux) [inline app/base/routes.py:37]: perhaps this
"Your email is already verified."

File: app/base/routes.py [hunk 3]
```diff
@@ -27,6 +28,52 @@ def route_errors(error):
...
+    from app.tasks import send_email
+
+    token = generate_confirmation_token(current_user.email)
+    confirm_url = url_for("base_blueprint.verify_email", token=token, _external=True)
+    template = render_template(
+        "verify_email.html",
+        url=confirm_url,
+        name=f"{current_user.first_name} {current_user.last_name}",
+    )
+    subject = "Verify Email"
```
Comment (sonlinux) [inline app/base/routes.py:61]: perhaps
subject = "Verify Email - Property Deals"

File: app/base/routes.py [hunk 4]
```diff
@@ -72,6 +121,19 @@ def register():
...
+        token = generate_confirmation_token(form.email.data)
+        confirm_url = url_for(
+            "base_blueprint.verify_email", token=token, _external=True
+        )
+        template = render_template(
+            "verify_email.html",
+            url=confirm_url,
+            name=f"{form.first_name.data} {form.last_name.data}",
+        )
+        subject = "Verify Email"
```
Comment (sonlinux) [inline app/base/routes.py:134]: perhaps
subject = "Verify Email - Property Deals"

File: app/base/routes.py [hunk 5]
```diff
@@ -72,6 +121,19 @@ def register():
...
+        send_email.delay(form.email.data, subject, template)
+        flash("A confirmation email has been sent", "success")
```
Comment (sonlinux) [inline app/base/routes.py:136]: `flash("A verification email has been sent", "success")`

File: app/base/utils.py [hunk 6]
```diff
@@ -2,12 +2,58 @@
...
 
 
+def generate_confirmation_token(user_email):
+    """
+    Generates a token and with user email as encoded data
+    """
+    serializer = URLSafeTimedSerializer(
+        os.environ.get("SECRET_KEY")
+    )
+    salt = os.environ.get("SECURITY_PASSWORD_SALT")
```
Comment (sonlinux) [inline app/base/utils.py:22]: seems like a good candidate for a util function

File: app/base/utils.py [hunk 7]
```diff
@@ -2,12 +2,58 @@
...
+
+def confirm_token(token, expiration=3600):
+    """
+    Decodes the data in the token, return False if token is not valid or has expired and return
+    decoded data if token is valid.
+    """
+    serializer = URLSafeTimedSerializer(
+        os.environ.get("SECRET_KEY")
+    )
+    salt = os.environ.get("SECURITY_PASSWORD_SALT")
```
Comment (sonlinux) [inline app/base/utils.py:35]: seems like a good candidate for a util function

File: app/base/utils.py [hunk 8]
```diff
@@ -31,7 +77,9 @@ def save_to_redis(image_file_list, username):
     img_dict = {}
     for file in image_file_list:
         _, file_ext = os.path.splitext(file.filename)  # Get file extension
-        new_img_name = uuid.uuid4().__str__()[:8]  # generate a random string to set as key for the image in img_dict
+        new_img_name = uuid.uuid4().__str__()[
+            :8
+        ]  # generate a random string to set as key for the image in img_dict
```
Comment (sonlinux) [inline app/base/utils.py:82]: lets reformat this by hand back to `new_img_name = uuid.uuid4().__str__()[:8] `

File: app/tasks.py [hunk 9]
```diff
@@ -35,7 +38,12 @@ def upload_to_s3(img_obj, s3_dir, folder_name):
             open(img_obj, "rb"),
             bucket,
             f"{s3_dir}{folder_name}/{img_obj}",
-            ExtraArgs={"ACL": "public-read", "ContentType": "image/jpeg" if file_ext in jpeg_extensions else "image/png"}
+            ExtraArgs={
+                "ACL": "public-read",
+                "ContentType": "image/jpeg"
+                if file_ext in jpeg_extensions
+                else "image/png",
```
Comment (sonlinux) [inline app/tasks.py:45]:
```
Contentype: image/jpeg
if file_ext in jpeg_extensions 
else Contentype: image/png
```

File: app/tasks.py [hunk 10]
```diff
@@ -56,3 +64,21 @@ def delete_img_obj(bucket, dir_to_del, image_list=None, filename=None):
...
+        to_emails=to,
+        subject=subject,
+        html_content=template,
+    )
+    try:
+        sg = SendGridAPIClient(os.environ.get("SENDGRID_API_KEY"))
+        response = sg.send(message)
+        print(response.status_code)
+        print(response.body)
+        print(response.headers)
```
Comment (sonlinux) [inline app/tasks.py:81]: We don't need this  in prod

File: app/tasks.py [hunk 11]
```diff
@@ -56,3 +64,21 @@ def delete_img_obj(bucket, dir_to_del, image_list=None, filename=None):
...
+        return "Email has been sent."
```
Comment (sonlinux) [inline app/tasks.py:82]: "Your verification email has been sent."

**Review by sonlinux — CHANGES_REQUESTED** [review]

File: app/base/utils.py (hunk 6 above)
Comment (Jachin-Manda) [reply app/base/utils.py:22]: I have declared them as global variables

**Review by sonlinux — APPROVED** [review]

**Pull request merged**